misc/python/materialize/mzbuild.py

```diff
@@ -671,6 +671,30 @@ def acquire(self, force_build: bool = False) -> None:
             else:
                 announce(f"Already built {spec}")
 
+    def push_tagged(self, tag: str, include_spec: bool = False) -> None:
+        """
+        Push all images in this DependencySet tagged with the given tag
+
+        All images must have been `acquired()` before calling this method. Only dependencies with
+        the `publish` bit set will be pushed.
+
+        Args:
+            tag: The Docker image tag to use
+            include_spec: If true, include the `mzbuild-<spec>` tag as part of the push
+        """
+        publishable = set(d.spec() for d in self if d.publish)
+        missing = publishable - set(docker_images())
+        assert not missing, "can only push images once they have been acquired"
+
+        for dep in self:
+            if dep.pushed() or not dep.publish:
```

Comment:
So I think the `dep.pushed()` in the condition will cause the wrong behavior, since we want to push the `unstable` tag irrespective of whether the corresponding mzbuild tag has been pushed.

Reply:
I think it'll be clearer and less error prone if we split `include_spec` into its own method, and make both of them check for whether the image has already been pushed to Docker Hub. We can also remove the `ResolvedImage.{push,pushed}` methods to centralize all the pushing logic here.

```diff
diff --git a/ci/test/build.py b/ci/test/build.py
index b9c89a827..4f67c5662 100644
--- a/ci/test/build.py
+++ b/ci/test/build.py
@@ -35,7 +35,8 @@ def main() -> None:
     commit_tag = f'unstable-{git.rev_parse("HEAD")}'
     deps = repo.resolve_dependencies(image for image in repo if image.publish)
     deps.acquire()
-    deps.push_tagged(commit_tag, include_spec=True)
+    deps.push()
+    deps.push_tagged(commit_tag)
 
     print("--- Staging Debian package")
     if os.environ["BUILDKITE_BRANCH"] == "main":
diff --git a/misc/python/materialize/mzbuild.py b/misc/python/materialize/mzbuild.py
index 952b322a7..c87bdbbb0 100644
--- a/misc/python/materialize/mzbuild.py
+++ b/misc/python/materialize/mzbuild.py
@@ -126,6 +126,19 @@ def docker_images() -> Set[str]:
         .split("\n")
     )
 
+def docker_image_pushed(name: str) -> bool:
+    """Check whether the named image is pushed to Docker Hub.
+
+    Note that this operation requires a rather slow network request.
+    """
+    proc = subprocess.run(
+        ["docker", "manifest", "inspect", name],
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+        env=dict(os.environ, DOCKER_CLI_EXPERIMENTAL="enabled"),
+    )
+    return proc.returncode == 0
+
 
 def chmod_x(path: Path) -> None:
     """Set the executable bit on a file or directory."""
@@ -517,27 +530,6 @@ class ResolvedImage:
         self.build()
         return AcquiredFrom.LOCAL_BUILD
 
-    def push(self) -> None:
-        """Push the image to Docker Hub.
-
-        The image is pushed under the Docker identifier returned by
-        `ResolvedImage.spec`.
-        """
-        spawn.runv(["docker", "push", self.spec()])
-
-    def pushed(self) -> bool:
-        """Check whether the image is pushed to Docker Hub.
-
-        Note that this operation requires a rather slow network request.
-        """
-        proc = subprocess.run(
-            ["docker", "manifest", "inspect", self.spec()],
-            stdout=subprocess.DEVNULL,
-            stderr=subprocess.DEVNULL,
-            env=dict(os.environ, DOCKER_CLI_EXPERIMENTAL="enabled"),
-        )
-        return proc.returncode == 0
-
     def run(self, args: List[str] = []) -> None:
         """Run a command in the image.
 
@@ -671,29 +663,24 @@ class DependencySet:
             else:
                 announce(f"Already built {spec}")
 
-    def push_tagged(self, tag: str, include_spec: bool = False) -> None:
-        """
-        Push all images in this DependencySet tagged with the given tag
-
-        All images must have been `acquired()` before calling this method. Only dependencies with
-        the `publish` bit set will be pushed.
+    def push(self) -> None:
+        """Push all publishable images in this dependency set to Docker Hub.
 
-        Args:
-            tag: The Docker image tag to use
-            include_spec: If true, include the `mzbuild-<spec>` tag as part of the push
+        Images are pushed using their spec as their tag. All images must have
+        been acquired (e.g., by `DependencySet.acquire`) before calling this
+        method.
         """
-        publishable = set(d.spec() for d in self if d.publish)
-        missing = publishable - set(docker_images())
-        assert not missing, "can only push images once they have been acquired"
+        for dep in self:
+            if dep.publish and not docker_image_pushed(dep.spec()):
+                spawn.runv(["docker", "push", dep.spec()])
 
+    def push_tagged(self, tag: str) -> None:
+        """Like push, but uses the specified tag instead of the image's spec."""
         for dep in self:
-            if dep.pushed() or not dep.publish:
-                continue
             name = dep.image.docker_name(tag)
-            spawn.runv(["docker", "tag", dep.spec(), name])
-            if include_spec:
-                dep.push()
-            spawn.runv(["docker", "push", name])
+            if dep.publish and not docker_image_pushed(name):
+                spawn.runv(["docker", "tag", dep.spec(), name])
+                spawn.runv(["docker", "push", name])
 
     def __iter__(self) -> Iterator[ResolvedImage]:
         return iter(self.dependencies.values())
```